### ISSUE TITLE
Memory leak in proxy-api

### ISSUE THREAD

**Issue by jon-walton** [linkerd/linkerd2#2382]
### What is the issue?

I woke up to alerts that one of the ingress controller pods was in crash loop backoff. It was the usual case of the ingress controller logs looking fine and then being killed due to failing a health check.

Looking further, I saw that the linkerd controller pods were using excessive amounts of memory (1.96GB, 535MB and 520MB). Linkerd was installed with `--ha --tls=optional --proxy-auto-inject`

Restarting the linkerd controllers and the failed nginx controller brought things back into a working state.

### How can it be reproduced?

Unknown

### Logs, error output, etc

The linkerd controller was logging this, roughly every second

```
linkerd linkerd-controller-5fcc8cb6fd-2mpxp linkerd-proxy 2019-02-26T00:17:06.540543684Z ERR! proxy={server=in listen=0.0.0.0:4143 remote=10.1.64.108:39444} linkerd2_proxy::proxy::http::router service error: in-flight limit exceeded
```

occasionally I would see this

```
linkerd linkerd-controller-5fcc8cb6fd-ndrth linkerd-proxy 2019-02-26T00:17:11.84895528Z WARN linkerd2_proxy::app::profiles error fetching profile for linkerd-proxy-api.linkerd.svc.cluster.local:8086: Inner(Upstream(Inner(Inner(Timeout(3s)))))
```

`linkerd endpoints` seemed fine

![image](https://user-images.githubusercontent.com/9859186/53384086-abc19d80-39b4-11e9-9a23-3a326183b98c.png)
Memory usage of 1 pod for the last 7 days

![image](https://user-images.githubusercontent.com/9859186/53384177-f6dbb080-39b4-11e9-9892-79627452f8b2.png)
The 3 proxy-apis were sitting at 127k, 28k, 27k goroutines. The sudden drop-off is when I restarted the pods

![image](https://user-images.githubusercontent.com/9859186/53384277-41f5c380-39b5-11e9-9e2f-24acdcd04752.png)
interestingly, we've only got a high number of open file descriptors in 1 of the controllers (again, proxy-api container). The others are sitting around 20ish

![image](https://user-images.githubusercontent.com/9859186/53384446-e7109c00-39b5-11e9-840f-b0eaded3fb6e.png)
ENOTCONN seems to show an increase

![image](https://user-images.githubusercontent.com/9859186/53384509-14f5e080-39b6-11e9-94b4-91b28a20c190.png)
The tooltip is showing a count of 2k in the 646k - 506k bucket?!?

![image](https://user-images.githubusercontent.com/9859186/53384694-c563e480-39b6-11e9-9cf8-962fb3ed7903.png)
proxy-api memstats

#### `linkerd check` output

I forgot to do `linkerd check` or `linkerd check --proxy` :(

### Environment

- Kubernetes Version: 1.13.3
- Cluster Environment: AKS-Engine 0.31.0
- Host OS:
- Linkerd version: stable-2.2.0

**Comment by olix0r** [comment]
Thanks for the great report @jon-walton. A few random questions...

About how many pods exist in your cluster? About how many of these were configured to use linkerd?

Were pods being created/deleted?

Had you created any service profiles?

**Comment by jon-walton** [comment]
`k -n ava get pod -l linkerd.io/control-plane-ns=linkerd --all-namespaces | wc -l`

adjusting for the linkerd pods, 16 pods are meshed, 95 pods exist in the cluster (including linkerd pods)

no pods were being created (the last pods created were 17 hours ago, waaay after the memory leak started

The only service profiles created are the ones with `linkerd install-sp`

**Comment by jon-walton** [comment]
@olix0r if my alerts are right, the ingress controller would have started to go into crash loop backoff around 4am (before the start of the charts unfortunately)

**Comment by dadjeibaah** [comment]
I ran some tests on an AKS cluster and it looks like we might be leaking connection from proxies to the destination service when getting service profiles. 

I ran a rough test with an nginx-ingress-controller and an injected proxy and a go client that made requests to the external IP of the cluster.  Each request the client made contained the `l5d-dst-override` header with a randomly generated string. There were a total of a 1000 unique strings used in this test.

When running each request, the first thing I observed was that I wasn't getting any response for each request. Only after the router hit its capacity did I start to see 503 responses.

```bash
Requesting host: YZUeIVtn
Requesting host: UkrjmTaT
Requesting host: BNQwjlpF
Requesting host: hBudrZXq
Requesting host: uEipSWNc
Requesting host: CqRLWuyI
Requesting host: CzrkZbMc
Received HTTP status: 503
Requesting host: ttbqeMhO
Received HTTP status: 503
```

Another thing I noticed was that the amount goroutines in the proxy-api kept rising with each successive test run and did not seem to go back down.

<img width="568" alt="Screen Shot 2019-03-11 at 2 55 05 PM" src="https://user-images.githubusercontent.com/2197104/54222093-f3b2eb00-44b1-11e9-9aa9-ebf8213a84a1.png">

Running a `pprof` profile on the proxy-api showed that a majority of those goroutines were stream requests to `GetProfile`

![Screen Shot 2019-03-12 at 10 34 04 AM](https://user-images.githubusercontent.com/2197104/54222314-6328da80-44b2-11e9-81fd-378216570059.png)

I then went to go look at the number of TCP connections that were open in proxy of the nginx-ingress-controller pod:
```bash
I have no name!@running-gecko-nginx-ingress-controller-75cb5fc57-2c962:/linkerd$ cat /proc/$(pgrep linkerd-proxy)/net/tcp | wc -l
1309
```

**Comment by gamer22026** [comment]
Had similar issue. Running 19.3.2 w/auto-inject no TLS. 66 meshed pods and 32 serviceprofiles. Memory usage was at 2.5G on the linkerd-controller and linkerd-proxy was throwing `linkerd2_proxy::app::profiles error fetching profile for someservice:443: Status { code: Unknown, message: "operation timed out after 3s" }` errors for services that had serviceprofiles applied and services were failing. Restarting the linkerd-controller and memory is back down to ~ 20M and everything started functioning again

**Comment by dadjeibaah** [comment]
@jon-walton @gamer22026, we just released [edge-19.3.3]( https://github.com/linkerd/linkerd2/releases/tag/edge-19.3.3) and I believe there is a fix for this issue in the release. Are you able to test this out to confirm that this is the case?

**Comment by siggy** [comment]
For reference, we believe this was fixed via linkerd/linkerd2-proxy#220.

**Comment by jon-walton** [comment]
@siggy @dadjeibaah  This seems to be fixed for me. thanks 👍
